tests/zfs-tests/tests/functional/deadman/deadman_ratelimit.ksh

```diff
@@ -0,0 +1,83 @@
+#!/bin/ksh -p
+#
+# CDDL HEADER START
+#
+# The contents of this file are subject to the terms of the
+# Common Development and Distribution License (the "License").
+# You may not use this file except in compliance with the License.
+#
+# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+# or http://www.opensolaris.org/os/licensing.
+# See the License for the specific language governing permissions
+# and limitations under the License.
+#
+# When distributing Covered Code, include this CDDL HEADER in each
+# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+# If applicable, add the following below this CDDL HEADER, with the
+# fields enclosed by brackets "[]" replaced with your own identifying
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+
+#
+# Portions Copyright 2021 iXsystems, Inc.
+#
+
+# DESCRIPTION:
+#	Verify spa deadman events are rate limited
+#
+# STRATEGY:
+#	1. Reduce the zfs_slow_io_events_per_second to 1.
+#	2. Reduce the zfs_deadman_ziotime_ms to 1ms.
+#	3. Write data to a pool and read it back.
+#	4. Verify events have been dropped.
+#
+
+. $STF_SUITE/include/libtest.shlib
+. $STF_SUITE/tests/functional/deadman/deadman.cfg
+
+verify_runnable "both"
+
+function cleanup
+{
+	default_cleanup_noexit
+
+	set_tunable64 SLOW_IO_EVENTS_PER_SECOND $OLD_SLOW_IO_EVENTS
+	set_tunable64 DEADMAN_ZIOTIME_MS $ZIOTIME_DEFAULT
+}
+
+log_assert "Verify spa deadman events are rate limited"
+log_onexit cleanup
+
+OLD_SLOW_IO_EVENTS=$(get_tunable SLOW_IO_EVENTS_PER_SECOND)
+log_must set_tunable64 SLOW_IO_EVENTS_PER_SECOND 1
+log_must set_tunable64 DEADMAN_ZIOTIME_MS 1
+
+# Create a new pool in order to use the updated deadman settings.
+default_setup_noexit $DISK1
+log_must zpool events -c
+
+mntpnt=$(get_prop mountpoint $TESTPOOL/$TESTFS)
+log_must file_write -b 1048576 -c 8 -o create -d 0 -f $mntpnt/file
+log_must zinject -a
+log_must dd if=$mntpnt/file of=$TEST_BASE_DIR/devnull oflag=sync
+
+fails=0
+TIMEOUT=60
+while ! eval "zpool events $TESTPOOL | grep -q dropped"; do
+	if (( fails++ > $TIMEOUT )); then
+		log_fail "No dropped events after $TIMEOUT seconds"
+	fi
+	sleep 1
+done
```

Comment:
This doesn't look right to me.  Why are we expecting dropped events here?  I'd expect to only see drops reported when we overflow the `zfs_zevent_len_max` queue depth and the ZED doesn't consume the event before it's discard.  Events which are internally ratelimited on the kernel side and never posted aren't considered to be drops.

What we've been meaning to do for a while is increase the default `zfs_zevent_len_max` to make drops far more uncommon.  The current value is really very low, and scaled of the number of cpus which doesn't make much sense.

Reply:
Thanks, I definitely misunderstood that. I'll look for another way to test this.

Reply:
In fact `zfs_is_ratelimiting_event()` does increment `ratelimit_dropped` via `fm_erpt_dropped_increment()` when the rate limiting kicks in, so I should expect to see dropped events. ZED is not running during these tests, either. I wonder if it's possible the zio is actually completing in under 1ms, so no deadman events are posted? That feels too good to be true.